### PR TITLE
Labeling

### PR DESCRIPTION
**Pull request by dholms**
Adds labeling to PDS - soon to be moved to AppView

Keyword labeling on text (which we aren't using yet) & automated img labeling using hive.

Return relevant labels on all views

### CODE REVIEW

File: packages/pds/src/app-view/services/indexing/plugins/follow.ts [hunk 1]
```diff
@@ -49,15 +48,17 @@ const findDuplicate = async (
...
   return [
-    messages.createNotification({
+    {
       userDid: obj.subjectDid,
       author: obj.creator,
       recordUri: obj.uri,
       recordCid: obj.cid,
-      reason: 'follow',
-    }),
+      reason: 'follow' as const,
```
Comment (bnewbold) [inline packages/pds/src/app-view/services/indexing/plugins/follow.ts:58]: While it is on my mind, I think I saw a user report of muting somebody, but still getting notifications if they follow/unfollow repeatedly.

File: packages/pds/src/api/app/bsky/notification/listNotifications.ts [hunk 2]
```diff
@@ -93,6 +98,7 @@ export default function (server: Server, ctx: AppContext) {
         record: common.cborBytesToRecord(notif.recordBytes),
         isRead: notif.indexedAt <= userState.lastSeenNotifs,
         indexedAt: notif.indexedAt,
+        labels: labels[notif.uri],
```
Comment (devinivy) [inline packages/pds/src/api/app/bsky/notification/listNotifications.ts:101]:
```suggestion
        labels: labels[notif.uri] ?? [],
```

File: packages/dev-env/src/mock/index.ts [hunk 3]
```diff
@@ -164,6 +165,61 @@ export async function generateMockSetup(env: DevEnv) {
     }
   }
 
+  // make some naught posts & label them
```
Comment (dholms) [inline packages/dev-env/src/mock/index.ts:168]:
```suggestion
  // make some naughty posts & label them
```

File: packages/dev-env/src/mock/index.ts [hunk 4]
```diff
@@ -164,6 +165,61 @@ export async function generateMockSetup(env: DevEnv) {
...
+          subjectCid: labeledPost.cid,
+          value: 'nudity',
+          negated: 0,
+          createdAt: new Date().toISOString(),
+        },
+        {
+          sourceDid: ctx.cfg.labelerDid,
+          subjectUri: filteredPost.uri,
+          subjectCid: filteredPost.cid,
+          value: 'csam',
```
Comment (dholms) [inline packages/dev-env/src/mock/index.ts:215]:
```suggestion
          value: 'dmca',
```

Comment (dholms) [reply packages/dev-env/src/mock/index.ts:215]: hey @pfrazee i changed this one cause it felt like a weird label in our dev-env 🙃

Comment (dholms) [reply packages/dev-env/src/mock/index.ts:215]: ah right good catch. i shouldve checked

File: packages/pds/src/config.ts [hunk 5]
```diff
@@ -120,6 +124,10 @@ export class ServerConfig {
     const emailNoReplyAddress =
       process.env.EMAIL_NO_REPLY_ADDRESS || 'noreply@blueskyweb.xyz'
 
+    const hiveApiKey = process.env.HIVE_API_KEY || undefined
```
Comment (devinivy) [inline packages/pds/src/config.ts:127]: Once we land this, should things be all good if we run the pds without configuring a hive API key, or is it required?

File: packages/pds/src/db/tables/label.ts [hunk 6]
```diff
@@ -0,0 +1,12 @@
+export const tableName = 'label'
+
+export interface Label {
+  sourceDid: string
+  subjectUri: string
+  subjectCid: string
+  value: string
+  negated: 0 | 1 // @TODO convert to boolean in app-view
```
Comment (devinivy) [inline packages/pds/src/db/tables/label.ts:8]: Is this todo still outstanding?

File: packages/pds/src/config.ts (hunk 5 above)
Comment (devinivy) [reply packages/pds/src/config.ts:127]: Oops, think I answered this for myself.

File: packages/pds/src/labeler/base.ts [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    const rows = labels.map((value) => ({
+      sourceDid: this.labelerDid,
+      subjectUri: uri.toString(),
+      subjectCid: cid.toString(),
+      value,
+      negated: 0 as const,
+      createdAt: new Date().toISOString(),
+    }))
+
+    await this.db.db.insertInto('label').values(rows).execute()
```
Comment (devinivy) [inline packages/pds/src/labeler/base.ts:49]: Do we need any handling for a record that has already been labeled?  E.g. maybe no-oping when trying to add a label that already exists with a `do nothing`?

File: packages/pds/src/labeler/hive.ts [hunk 8]
```diff
@@ -0,0 +1,118 @@
+import stream from 'stream'
+import axios from 'axios'
+import FormData from 'form-data'
+import { Labeler } from './base'
+import Database from '../db'
+import { BlobStore } from '@atproto/repo'
+import { keywordLabeling } from './util'
+
+const HIVE_ENDPOINT = 'https://api.thehive.ai/api/v2/task/sync'
```
Comment (devinivy) [inline packages/pds/src/labeler/hive.ts:9]: Do they have a sandbox environment we should be using in dev?

File: packages/pds/src/labeler/util.ts [hunk 9]
```diff
@@ -0,0 +1,107 @@
+import * as lex from '../lexicon/lexicons'
+import { Record as PostRecord } from '../lexicon/types/app/bsky/feed/post'
+import { Record as ProfileRecord } from '../lexicon/types/app/bsky/actor/profile'
+import { isMain as isEmbedImage } from '../lexicon/types/app/bsky/embed/images'
+import { isMain as isEmbedExternal } from '../lexicon/types/app/bsky/embed/external'
+import { isMain as isEmbedRecordWithMedia } from '../lexicon/types/app/bsky/embed/recordWithMedia'
+import { CID } from 'multiformats/cid'
```
Comment (devinivy) [inline packages/pds/src/labeler/util.ts:7]: Just ignore me haha
```suggestion
import { CID } from 'multiformats/cid'
import * as lex from '../lexicon/lexicons'
import { Record as PostRecord } from '../lexicon/types/app/bsky/feed/post'
import { Record as ProfileRecord } from '../lexicon/types/app/bsky/actor/profile'
import { isMain as isEmbedImage } from '../lexicon/types/app/bsky/embed/images'
import { isMain as isEmbedExternal } from '../lexicon/types/app/bsky/embed/external'
import { isMain as isEmbedRecordWithMedia } from '../lexicon/types/app/bsky/embed/recordWithMedia'
```

File: packages/pds/src/labeler/util.ts [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+  }
+}
+
+export const keywordLabeling = (
+  keywords: Record<string, string>,
+  text: string,
+): string[] => {
+  const labels: string[] = []
+  for (const word of Object.keys(keywords)) {
+    if (text.includes(word)) {
```
Comment (devinivy) [inline packages/pds/src/labeler/util.ts:92]: This is pretty loose, e.g. could match a bad word that is a part of a good word.  Also not sure if I see normalization happening anywhere, e.g. removing accents or at least upper/lower-casing.

File: packages/pds/tests/labeler/hive.test.ts [hunk 11]
```diff
@@ -0,0 +1,16 @@
+import fs from 'fs/promises'
+import * as hive from '../../src/labeler/hive'
+
+describe('labeling', () => {
+  it('correclty parses hive responses', async () => {
```
Comment (devinivy) [inline packages/pds/tests/labeler/hive.test.ts:5]:
```suggestion
  it('correctly parses hive responses', async () => {
```

**Review by devinivy — APPROVED** [review]
Lookin real solid—a few minor questions that we can work through any time.

File: packages/pds/src/config.ts (hunk 5 above)
Comment (dholms) [reply packages/pds/src/config.ts:127]: yup if it's not included, it'll just fallback to keyword labeler

File: packages/pds/src/db/tables/label.ts (hunk 6 above)
Comment (dholms) [reply packages/pds/src/db/tables/label.ts:8]: it's for when we put this on the AppView & we're using postgres which allows for booleans. using int2 here cuz sqlite

File: packages/pds/src/labeler/base.ts (hunk 7 above)
Comment (dholms) [reply packages/pds/src/labeler/base.ts:49]: oh great call! we need an "on conflict" here

File: packages/pds/src/labeler/hive.ts (hunk 8 above)
Comment (dholms) [reply packages/pds/src/labeler/hive.ts:9]: don't think so :/

File: packages/pds/src/labeler/util.ts (hunk 10 above)
Comment (dholms) [reply packages/pds/src/labeler/util.ts:92]: ah yup we should normalize on lower-casing

keyword labeling is going to be very crude, it's just the simplest possible auto-labeling. I don't think we're starting off with it (& idk if we'll use it much at all actually). We can tighten it up if we think it's going to start to become load bearing

**Review by devinivy — APPROVED** [review]

**Pull request merged**